lib/parsers/package-lock-parser.ts

```diff
@@ -49,66 +64,240 @@ export class PackageLockParser implements LockfileParser {
       version: manifestFile.version || '',
     };
 
-    const topLevelDeps: Dep[] = getTopLevelDeps(manifestFile, includeDev);
-
     // asked to process empty deps
     if (_.isEmpty(manifestFile.dependencies) && !includeDev) {
       return depTree;
     }
 
-    await Promise.all(topLevelDeps.map(async (dep) => {
-      depTree.dependencies[dep.name] = await this.buildSubTreeRecursiveFromPackageLock(
-        dep.name, ['dependencies'], packageLock, [], dep.dev);
-    }));
+    // prepare a flat map, where dependency path is a key to dependency object
+    // path is an unique identifier for each dependency and corresponds to the
+    // relative path on disc
+    const depMap: DepMap = this.flattenLockfile(packageLock);
 
+    // all paths are identified, we can create a graph representing what depends on what
+    const depGraph: graphlib.Graph = this.createGraphOfDependencies(depMap);
+
+    // topological sort will be applied and it requires acyclic graphs
+    let cycleStarts = {}; // cycle starts are need for top level dependencies
+    if (!graphlib.alg.isAcyclic(depGraph)) {
+      const cycles: string[][] = graphlib.alg.findCycles(depGraph);
+      for (const cycle of cycles) {
+        cycleStarts = {...cycleStarts, ...this.removeCycle(cycle, depMap, depGraph)};
+      }
+    }
+
+    // transform depMap to a map of PkgTrees
+    const depTrees: {[depPath: string]: PkgTree} = await this.createDepTrees(depMap, depGraph);
+
+    // get trees for dependencies from manifest file
+    const topLevelDeps: Dep[] = getTopLevelDeps(manifestFile, includeDev);
+    for (const dep of topLevelDeps) {
+      // if any of top level dependencies is a part of cycle
+      // it now has a different item in the map
+      const depName = cycleStarts[dep.name] ? cycleStarts[dep.name] : dep.name;
+      if (depTrees[depName]) {
+        // if the top level dependency is dev, all children are dev
+        depTree.dependencies[dep.name] = dep.dev ?
+          this.setDevDepRec(_.cloneDeep(depTrees[depName])) : depTrees[depName];
+      }
+    }
```

Comment:
would add the `seImmediatePromise()` call in this loop too, as the `setDevDepRec` might be blocking (maybe call it only if `dep.dev`)